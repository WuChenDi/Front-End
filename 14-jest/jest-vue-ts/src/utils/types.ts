import { PropType } from "vue";

export const HeaderPropsDefine = {
  add: {
    type: Function as PropType<(v: any) => void>,
    required: true,
  },
} as const;

export const UnDoListPropsDefine = {
  list: {
    type: Object as PropType<UnDoListType[]>,
    required: true,
  },
  delete: {
    type: Function as PropType<(v: any) => void>,
    required: true,
  },
  status: {
    type: Function as PropType<(v: any) => void>,
    required: true,
  },
  reset: {
    type: Function as PropType<() => void>,
    required: true,
  },
  change: {
    type: Function as PropType<(v: any) => void>,
    required: true,
  },
} as const;

export interface UnDoListType {
  status: string;
  value: string;
}

export interface UnDoChangeParameter {
  value: string;
  index: number;
}
