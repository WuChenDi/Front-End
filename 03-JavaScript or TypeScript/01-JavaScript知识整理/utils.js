const localStorage = window.localStorage;
const sessionStorage = window.sessionStorage;

// 设置localStorage存储数据
export const setLocal = (name, content) => {
	if (!name) return;
	if (typeof content !== "string") {
		content = JSON.stringify(content);
	}
	localStorage.setItem(name, content);
};

// 获取localStorage存储数据
export const getLocal = (name) => {
	if (!name) return;
	let value = localStorage.getItem(name);
	if (value !== null) {
		try {
			value = JSON.parse(value);
		} catch (error) {
			value = value;
		}
	}
	return value;
};

// 删除localStorage存储数据
export const removeLocal = (name) => {
	if (!name) return;
	localStorage.removeItem(name);
};

// 设置sessionStorage存储数据
export const setSession = (name, content) => {
	if (!name) return;
	if (typeof content !== "string") {
		content = JSON.stringify(content);
	}
	sessionStorage.setItem(name, content);
};

// 获取sessionStorage存储数据
export const getSession = (name) => {
	if (!name) return;
	let value = sessionStorage.getItem(name);
	if (value !== null) {
		try {
			value = JSON.parse(value);
		} catch (error) {
			value = value;
		}
	}
	return value;
};

// 删除sessionStorage存储数据
export const removeSession = (name) => {
	if (!name) return;
	sessionStorage.removeItem(name);
};

/**
 * @method 获取URL的中参数
 * @param {*} strUrl
 */
export const getQueryString = (strUrl) => {
	let reg = new RegExp("(^|&)" + strUrl + "=([^&]*)(&|$)", "i");
	let r = window.location.search.substr(1).match(reg); //获取url中"?"符后的字符串并正则匹配
	let context = "";
	if (r != null) context = r[2];
	reg = null;
	r = null;
	return context == null || context == "" || context == "undefined"
		? ""
		: context;
};

/**
 * @method 时间格式化
 * @param {string} time
 * @param {string} format yyyy/mm/dd hh:ii:ss(2019/07/24 09:45:43) yy/m/d hh:ii:ss(19/07/24 09:45:43) yyyy/mm/dd w(2019/07/24 星期三) mm/dd/yyyy(07/24/2019)
 * @returns
 */
export const formatTime = (time, format = "yyyy-mm-dd") => {
	const d = time ? new Date(time) : new Date();
	const t = (i) => {
		return (i < 10 ? "0" : "") + i;
	};

	const year = d.getFullYear();
	const month = d.getMonth() + 1;
	const day = d.getDate();
	const hour = d.getHours();
	const minutes = d.getMinutes();
	const seconds = d.getSeconds();
	const weekday = d.getDay();

	return format.replace(
		/(yy){1,2}|m{1,2}|d{1,2}|h{1,2}|i{1,2}|s{1,2}|w{1,2}/gi,
		function (r) {
			switch (r.toUpperCase()) {
				case "YY":
					return ("" + year).substr(2);
				case "YYYY":
					return year;
				case "M":
					return month;
				case "MM":
					return t(month);
				case "D":
					return day;
				case "DD":
					return t(day);
				case "H":
					return hour;
				case "HH":
					return t(hour);
				case "I":
					return minutes;
				case "II":
					return t(minutes);
				case "S":
					return seconds;
				case "SS":
					return t(seconds);
				case "W":
					return `星期${["日", "一", "二", "三", "四", "五", "六"][weekday]}`;
				case "WW":
					return [
						"Sunday",
						"Monday",
						"TuesDay",
						"Wednesday",
						"Thursday",
						"Friday",
						"Saturday",
					][weekday];
			}
		}
	);
};

/**
 * @description 函数防抖
 * @param {Function} method 延时调用函数
 * @param {Number} wait 延迟时长
 * @param {Boolean} immediate 立即执行选项
 */
export const Ddebounce = (method, wait, immediate) => {
	if (typeof method != "function") {
		throw new TypeError("Expected a function");
	}
	let timeout;
	// Ddebounce函数为返回值
	// 使用Async/Await处理异步，如果函数异步执行，等待setTimeout执行完，拿到原函数返回值后将其返回
	// args为返回函数调用时传入的参数，传给method
	let Ddebounce = function (...args) {
		return new Promise((resolve) => {
			// 用于记录原函数执行结果
			let result;
			// 将method执行时this的指向设为 debounce 返回的函数被调用时的this指向
			let context = this;
			// 如果存在定时器则将其清除
			if (timeout) {
				clearTimeout(timeout);
			}
			// 立即执行需要两个条件，一是immediate为true，二是timeout未被赋值或被置为null
			if (immediate) {
				// 如果定时器不存在，则立即执行，并设置一个定时器，wait毫秒后将定时器置为null
				// 这样确保立即执行后wait毫秒内不会被再次触发
				let callNow = !timeout;
				timeout = setTimeout(() => {
					timeout = null;
				}, wait);
				// 如果满足上述两个条件，则立即执行并记录其执行结果
				if (callNow) {
					result = method.apply(context, args);
					resolve(result);
				}
			} else {
				// 如果immediate为false，则等待函数执行并记录其执行结果
				// 并将Promise状态置为fullfilled，以使函数继续执行
				timeout = setTimeout(() => {
					// args是一个数组，所以使用fn.apply
					// 也可写作method.call(context, ...args)
					result = method.apply(context, args);
					resolve(result);
				}, wait);
			}
		});
	};

	// 在返回的 Ddebounce 函数上添加取消方法
	Ddebounce.cancel = function () {
		clearTimeout(timeout);
		timeout = null;
	};
	return Ddebounce;
};

/**
 * @method 节流
 * @param {*} func
 * @param {*} delay
 */
export const Dthrottle = (func, delay) => {
	let timer = null;
	return function () {
		let context = this;
		let args = arguments;
		if (!timer) {
			timer = setTimeout(function () {
				func.apply(context, args);
				timer = null;
			}, delay);
		}
	};
};

/**
 * @method 数组扁平化
 * @param {*} arry
 */
export const deepFlatten = (arr) => {
	return arr
		.toString()
		.split(",")
		.map((item) => +item);
	// or
	// return [].concat(...arr.map(item => (Array.isArray(item) ? deepFlatten(item) : item)));
};

/**
 * @method 判断类型
 * @param {*} obj
 * @returns
 */
export const toType = (obj) => {
	return {}.toString
		.call(obj)
		.match(/\s([a-zA-Z]+)/)[1]
		.toLowerCase();
};

// html 转义
export const escapeHtml = (str) => {
	if (!str) return;
	str = str.replace(/&/g, "&amp;");
	str = str.replace(/</g, "&lt;");
	str = str.replace(/>/g, "&gt;");
	str = str.replace(/“/g, "&quto;");
	str = str.replace(/'/g, "&#39;");
	return str;
};
