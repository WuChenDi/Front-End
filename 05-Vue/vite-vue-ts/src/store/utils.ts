import type { modules } from './modules'

/* 获取到 getters 结构类型 */
// 匹配到 单个module 下的 getter，小技巧 infer 为某一项
type GetGetter<GetterType> = GetterType extends { getters: infer G } ? G : unknown
// 获取 vuex 所有的 getters 模块
type GetGetters<GetterTypes> = {
  [K in keyof GetterTypes]: GetGetter<GetterTypes[K]>
}
type ModuleGetters = GetGetters<typeof modules>

// --------------------

/* 获取到 mutations 结构类型 */
// 配到 单个 module 下的 mutations
type GetMutation<MutationType> = MutationType extends { mutations: infer M } ? M : unknown
// 获取 vuex 所有的 mutations 模块
type GetMutations<MutationTypes> = {
  [K in keyof MutationTypes]: GetMutation<MutationTypes[K]>
}
type ModuleMutations = GetMutations<typeof modules>

// --------------------

/* 获取到 actions 结构类型 */
// 配到 单个 module 下的 action
type GetAction<ActionType> = ActionType extends { actions: infer A } ? A : unknown
// 获取 vuex 所有的 actions 模块
type GetActions<ActionTypes> = {
  [K in keyof ActionTypes]: GetAction<ActionTypes[K]>
}
type ModuleActions = GetActions<typeof modules>

// --------------------

/* Getter/Commit/Dispatch 智能提示处理 */
// gettters[模块名/方法]、commit[模块名/方法]、dispatch[模块名/方法]
// 传入的是 keyof 有可能是 symbol | number,所以 Prefix & string 取其中的string
type AddPrefix<Prefix, Keys> = `${Prefix & string}/${Keys & string}`

// 调换一下顺序：user: "user/isLogin" => "user/isLogin": user
type GetSpliceKey<P, Module> = AddPrefix<P, keyof Module>

type GetSpliceKeys<Modules> = {
  [K in keyof Modules]: GetSpliceKey<K, Modules[K]>
}[keyof Modules]
// type xx = GetSpliceKeys<ModuleGetters>

type GetFunc<T, A, B> = T[A & keyof T][B & keyof T[A & keyof T]]
type GetSpliceObj<T> = {
  // K extends `${infer A}/${infer B}`相当于  user/isLogin  A=>user B=>isLogin
  [K in GetSpliceKeys<T>]: K extends `${infer A}/${infer B}` ? GetFunc<T, A, B> : unknown
}

// --------------------

/* Getters/Mutations/Actons 拼接好 xxx/xxx 的格式  */
type GenGetters = GetSpliceObj<ModuleGetters>
type Getters = {
  [K in keyof GenGetters]: ReturnType<GenGetters[K]>
}

// --------------------

type GenMutations = GetSpliceObj<ModuleMutations>
type Mutations = {
  [K in keyof GenMutations]: ReturnType<GenMutations[K]>
}

// --------------------

type GenActions = GetSpliceObj<ModuleActions>
type Actions = {
  [K in keyof GenActions]: ReturnType<GenActions[K]>
}

// --------------------

// commit 获取 payload 参数类型
type MutationsPayload = {
  // Parameters 获取函数参数
  [K in keyof GenMutations]: Parameters<GenMutations[K]>[1]
}

interface GetCommit<T> {
  // 可选参数类型，会自动加上undefined
  <K extends keyof T>(mutation: K, payload?: T[K]): void
}

type Commit = GetCommit<MutationsPayload>

// --------------------

// dispatch 获取 payload 参数类型
type ActionPayload = {
  // Parameters 获取函数参数
  [K in keyof GenActions]: Parameters<GenActions[K]>[1]
}
interface GetDispatch<T> {
  // 可选参数类型，会自动加上undefined
  <K extends keyof T>(action: K, payload?: T[K]): Promise<unknown>
}

type Dispatch = GetDispatch<ActionPayload>

// --------------------

export type { Getters, Mutations, Actions, Dispatch, Commit }
