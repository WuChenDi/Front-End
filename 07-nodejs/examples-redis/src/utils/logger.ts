/* eslint-disable @typescript-eslint/no-unsafe-argument */
/* eslint-disable no-console */

const getTimestamp = () => {
  return new Date().toLocaleString() + '.' + new Date().getMilliseconds()
}

const logWithTimestamp = (message?: any, ...optionalParams: any[]) => {
  console.log(`[${getTimestamp()}]`, message, ...optionalParams)
}

const infoWithTimestamp = (message?: any, ...optionalParams: any[]) => {
  console.info(`[${getTimestamp()}]`, message, ...optionalParams)
}

const warnWithTimestamp = (message?: any, ...optionalParams: any[]) => {
  console.warn(`[${getTimestamp()}]`, message, ...optionalParams)
}

const errorWithTimestamp = (message?: any, ...optionalParams: any[]) => {
  console.error(`[${getTimestamp()}]`, message, ...optionalParams)
}

const logger = {
  log: logWithTimestamp,
  info: infoWithTimestamp,
  warn: warnWithTimestamp,
  error: errorWithTimestamp,
}

export default logger
