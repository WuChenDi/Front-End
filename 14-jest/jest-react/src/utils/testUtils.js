export const findTestWrapper = (wrapper, tag) => {
  return wrapper.find(`[data-test="${tag}"]`);
};
