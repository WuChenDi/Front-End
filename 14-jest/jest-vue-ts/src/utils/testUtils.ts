export const findTestWrapper = (wrapper: any, tag: string) => {
  return wrapper.find(`[data-test = "${tag}"]`);
};
