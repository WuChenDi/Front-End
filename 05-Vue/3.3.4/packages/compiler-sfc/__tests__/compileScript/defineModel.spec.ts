import { BindingTypes } from '@vue/compiler-core'
import { compileSFCScript as compile, assertCode } from '../utils'

describe('defineModel()', () => {
  test('basic usage', () => {
    const { content, bindings } = compile(
      `
      <script setup>
      const modelValue = defineModel({ required: true })
      const c = defineModel('count')
      </script>
      `,
      { defineModel: true }
    )
    assertCode(content)
    expect(content).toMatch('props: {')
    expect(content).toMatch('"modelValue": { required: true },')
    expect(content).toMatch('"count": {},')
    expect(content).toMatch('emits: ["update:modelValue", "update:count"],')
    expect(content).toMatch(
      `const modelValue = _useModel(__props, "modelValue")`
    )
    expect(content).toMatch(`const c = _useModel(__props, "count")`)
    expect(content).toMatch(`return { modelValue, c }`)
    expect(content).not.toMatch('defineModel')

    expect(bindings).toStrictEqual({
      modelValue: BindingTypes.SETUP_REF,
      count: BindingTypes.PROPS,
      c: BindingTypes.SETUP_REF
    })
  })

  test('w/ defineProps and defineEmits', () => {
    const { content, bindings } = compile(
      `
      <script setup>
      defineProps({ foo: String })
      defineEmits(['change'])
      const count = defineModel({ default: 0 })
      </script>
    `,
      { defineModel: true }
    )
    assertCode(content)
    expect(content).toMatch(`props: _mergeModels({ foo: String }`)
    expect(content).toMatch(`"modelValue": { default: 0 }`)
    expect(content).toMatch(`const count = _useModel(__props, "modelValue")`)
    expect(content).not.toMatch('defineModel')
    expect(bindings).toStrictEqual({
      count: BindingTypes.SETUP_REF,
      foo: BindingTypes.PROPS,
      modelValue: BindingTypes.PROPS
    })
  })

  test('w/ array props', () => {
    const { content, bindings } = compile(
      `
      <script setup>
      defineProps(['foo', 'bar'])
      const count = defineModel('count')
      </script>
    `,
      { defineModel: true }
    )
    assertCode(content)
    expect(content).toMatch(`props: _mergeModels(['foo', 'bar'], {
    "count": {},
  })`)
    expect(content).toMatch(`const count = _useModel(__props, "count")`)
    expect(content).not.toMatch('defineModel')
    expect(bindings).toStrictEqual({
      foo: BindingTypes.PROPS,
      bar: BindingTypes.PROPS,
      count: BindingTypes.SETUP_REF
    })
  })

  test('w/ local flag', () => {
    const { content } = compile(
      `<script setup>
      const foo = defineModel({ local: true, default: 1 })
      const bar = defineModel('bar', { [key]: true })
      const baz = defineModel('baz', { ...x })
      const qux = defineModel('qux', x)

      const foo2 = defineModel('foo2', { local: true, ...x })

      const local = true
      const hoist = defineModel('hoist', { local })
      </script>`,
      { defineModel: true }
    )
    assertCode(content)
    expect(content).toMatch(`_useModel(__props, "modelValue", { local: true })`)
    expect(content).toMatch(`_useModel(__props, "bar", { [key]: true })`)
    expect(content).toMatch(`_useModel(__props, "baz", { ...x })`)
    expect(content).toMatch(`_useModel(__props, "qux", x)`)
    expect(content).toMatch(`_useModel(__props, "foo2", { local: true })`)
    expect(content).toMatch(`_useModel(__props, "hoist", { local })`)
  })

  test('w/ types, basic usage', () => {
    const { content, bindings } = compile(
      `
      <script setup lang="ts">
      const modelValue = defineModel<boolean | string>()
      const count = defineModel<number>('count')
      const disabled = defineModel<number>('disabled', { required: false })
      const any = defineModel<any | boolean>('any')
      </script>
      `,
      { defineModel: true }
    )
    assertCode(content)
    expect(content).toMatch('"modelValue": { type: [Boolean, String] }')
    expect(content).toMatch('"count": { type: Number }')
    expect(content).toMatch(
      '"disabled": { type: Number, ...{ required: false } }'
    )
    expect(content).toMatch('"any": { type: Boolean, skipCheck: true }')
    expect(content).toMatch(
      'emits: ["update:modelValue", "update:count", "update:disabled", "update:any"]'
    )

    expect(content).toMatch(
      `const modelValue = _useModel(__props, "modelValue")`
    )
    expect(content).toMatch(`const count = _useModel(__props, "count")`)
    expect(content).toMatch(`const disabled = _useModel(__props, "disabled")`)
    expect(content).toMatch(`const any = _useModel(__props, "any")`)

    expect(bindings).toStrictEqual({
      modelValue: BindingTypes.SETUP_REF,
      count: BindingTypes.SETUP_REF,
      disabled: BindingTypes.SETUP_REF,
      any: BindingTypes.SETUP_REF
    })
  })

  test('w/ types, production mode', () => {
    const { content, bindings } = compile(
      `
      <script setup lang="ts">
      const modelValue = defineModel<boolean>()
      const fn = defineModel<() => void>('fn')
      const fnWithDefault = defineModel<() => void>('fnWithDefault', { default: () => null })
      const str = defineModel<string>('str')
      const optional = defineModel<string>('optional', { required: false })
      </script>
      `,
      { defineModel: true, isProd: true }
    )
    assertCode(content)
    expect(content).toMatch('"modelValue": { type: Boolean }')
    expect(content).toMatch('"fn": {}')
    expect(content).toMatch(
      '"fnWithDefault": { type: Function, ...{ default: () => null } },'
    )
    expect(content).toMatch('"str": {}')
    expect(content).toMatch('"optional": { required: false }')
    expect(content).toMatch(
      'emits: ["update:modelValue", "update:fn", "update:fnWithDefault", "update:str", "update:optional"]'
    )
    expect(content).toMatch(
      `const modelValue = _useModel(__props, "modelValue")`
    )
    expect(content).toMatch(`const fn = _useModel(__props, "fn")`)
    expect(content).toMatch(`const str = _useModel(__props, "str")`)
    expect(bindings).toStrictEqual({
      modelValue: BindingTypes.SETUP_REF,
      fn: BindingTypes.SETUP_REF,
      fnWithDefault: BindingTypes.SETUP_REF,
      str: BindingTypes.SETUP_REF,
      optional: BindingTypes.SETUP_REF
    })
  })
})
