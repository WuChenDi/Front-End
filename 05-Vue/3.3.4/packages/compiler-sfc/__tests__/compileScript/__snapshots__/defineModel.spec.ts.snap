// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`defineModel() > basic usage 1`] = `
"import { useModel as _useModel } from 'vue'

export default {
  props: {
    \\"modelValue\\": { required: true },
    \\"count\\": {},
  },
  emits: [\\"update:modelValue\\", \\"update:count\\"],
  setup(__props, { expose: __expose }) {
  __expose();

      const modelValue = _useModel(__props, \\"modelValue\\")
      const c = _useModel(__props, \\"count\\")
      
return { modelValue, c }
}

}"
`;

exports[`defineModel() > w/ array props 1`] = `
"import { useModel as _useModel, mergeModels as _mergeModels } from 'vue'

export default {
  props: _mergeModels(['foo', 'bar'], {
    \\"count\\": {},
  }),
  emits: [\\"update:count\\"],
  setup(__props, { expose: __expose }) {
  __expose();

      
      const count = _useModel(__props, \\"count\\")
      
return { count }
}

}"
`;

exports[`defineModel() > w/ defineProps and defineEmits 1`] = `
"import { useModel as _useModel, mergeModels as _mergeModels } from 'vue'

export default {
  props: _mergeModels({ foo: String }, {
    \\"modelValue\\": { default: 0 },
  }),
  emits: _mergeModels(['change'], [\\"update:modelValue\\"]),
  setup(__props, { expose: __expose }) {
  __expose();

      
      
      const count = _useModel(__props, \\"modelValue\\")
      
return { count }
}

}"
`;

exports[`defineModel() > w/ local flag 1`] = `
"import { useModel as _useModel } from 'vue'
const local = true
      
export default {
  props: {
    \\"modelValue\\": { local: true, default: 1 },
    \\"bar\\": { [key]: true },
    \\"baz\\": { ...x },
    \\"qux\\": x,
    \\"foo2\\": { local: true, ...x },
    \\"hoist\\": { local },
  },
  emits: [\\"update:modelValue\\", \\"update:bar\\", \\"update:baz\\", \\"update:qux\\", \\"update:foo2\\", \\"update:hoist\\"],
  setup(__props, { expose: __expose }) {
  __expose();

      const foo = _useModel(__props, \\"modelValue\\", { local: true })
      const bar = _useModel(__props, \\"bar\\", { [key]: true })
      const baz = _useModel(__props, \\"baz\\", { ...x })
      const qux = _useModel(__props, \\"qux\\", x)

      const foo2 = _useModel(__props, \\"foo2\\", { local: true })

      const hoist = _useModel(__props, \\"hoist\\", { local })
      
return { foo, bar, baz, qux, foo2, local, hoist }
}

}"
`;

exports[`defineModel() > w/ types, basic usage 1`] = `
"import { useModel as _useModel, defineComponent as _defineComponent } from 'vue'

export default /*#__PURE__*/_defineComponent({
  props: {
    \\"modelValue\\": { type: [Boolean, String] },
    \\"count\\": { type: Number },
    \\"disabled\\": { type: Number, ...{ required: false } },
    \\"any\\": { type: Boolean, skipCheck: true },
  },
  emits: [\\"update:modelValue\\", \\"update:count\\", \\"update:disabled\\", \\"update:any\\"],
  setup(__props, { expose: __expose }) {
  __expose();

      const modelValue = _useModel(__props, \\"modelValue\\")
      const count = _useModel(__props, \\"count\\")
      const disabled = _useModel(__props, \\"disabled\\")
      const any = _useModel(__props, \\"any\\")
      
return { modelValue, count, disabled, any }
}

})"
`;

exports[`defineModel() > w/ types, production mode 1`] = `
"import { useModel as _useModel, defineComponent as _defineComponent } from 'vue'

export default /*#__PURE__*/_defineComponent({
  props: {
    \\"modelValue\\": { type: Boolean },
    \\"fn\\": {},
    \\"fnWithDefault\\": { type: Function, ...{ default: () => null } },
    \\"str\\": {},
    \\"optional\\": { required: false },
  },
  emits: [\\"update:modelValue\\", \\"update:fn\\", \\"update:fnWithDefault\\", \\"update:str\\", \\"update:optional\\"],
  setup(__props, { expose: __expose }) {
  __expose();

      const modelValue = _useModel(__props, \\"modelValue\\")
      const fn = _useModel(__props, \\"fn\\")
      const fnWithDefault = _useModel(__props, \\"fnWithDefault\\")
      const str = _useModel(__props, \\"str\\")
      const optional = _useModel(__props, \\"optional\\")
      
return { modelValue, fn, fnWithDefault, str, optional }
}

})"
`;
